File: edit_server.py
#!/usr/bin/env python3
import html
import json
import os
import re
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer


ROOT = os.path.abspath(os.path.dirname(__file__))
DEFAULT_PORT = 8787


def _safe_join(root: str, rel: str) -> str:
    rel = rel.lstrip("/").replace("\\", "/")
    if ".." in rel.split("/"):
        raise ValueError("bad path")
    out = os.path.abspath(os.path.join(root, rel))
    if not out.startswith(root + os.sep):
        raise ValueError("bad path")
    return out


def _replace_edit_id(html_text: str, edit_id: str, new_text: str) -> str:
    # Only replaces the *inner content* of a tag that has data-edit-id="...".
    # We store plain text (escaped) so it stays safe and predictable.
    escaped = html.escape(new_text, quote=False)

    pattern = re.compile(
        r'(<(?P<tag>[a-zA-Z0-9]+)(?P<attrs>[^>]*\sdata-edit-id="' + re.escape(edit_id) + r'"[^>]*)>)'
        r'(?P<inner>[\s\S]*?)'
        r'(</(?P=tag)>)',
        re.MULTILINE,
    )

    match = pattern.search(html_text)
    if not match:
        raise KeyError(f"edit id not found: {edit_id}")

    start, end = match.span("inner")
    return html_text[:start] + escaped + html_text[end:]


class Handler(BaseHTTPRequestHandler):
    def _send(self, status: int, body: bytes, content_type: str = "text/plain; charset=utf-8"):
        self.send_response(status)
        self.send_header("Content-Type", content_type)
        self.send_header("Content-Length", str(len(body)))
        self.end_headers()
        self.wfile.write(body)

    def do_GET(self):
        path = self.path.split("?", 1)[0]
        if path == "/":
            path = "/scenarios.html"
        try:
            file_path = _safe_join(ROOT, path)
            if not os.path.isfile(file_path):
                self._send(404, b"Not found")
                return
            with open(file_path, "rb") as f:
                data = f.read()
            ctype = "text/html; charset=utf-8"
            if file_path.endswith(".css"):
                ctype = "text/css; charset=utf-8"
            elif file_path.endswith(".js"):
                ctype = "text/javascript; charset=utf-8"
            elif file_path.endswith(".png"):
                ctype = "image/png"
            elif file_path.endswith(".jpg") or file_path.endswith(".jpeg"):
                ctype = "image/jpeg"
            elif file_path.endswith(".webp"):
                ctype = "image/webp"
            self._send(200, data, ctype)
        except Exception as e:
            self._send(500, str(e).encode("utf-8"))

    def do_POST(self):
        if self.path.split("?", 1)[0] != "/__save":
            self._send(404, b"Not found")
            return

        try:
            length = int(self.headers.get("Content-Length", "0"))
            raw = self.rfile.read(length)
            payload = json.loads(raw.decode("utf-8"))

            rel_file = payload.get("file")
            edit_id = payload.get("id")
            text = payload.get("text", "")
            if not rel_file or not edit_id:
                self._send(400, b"Missing file or id")
                return

            target = _safe_join(ROOT, rel_file)
            if not os.path.isfile(target):
                self._send(404, b"File not found")
                return

            with open(target, "r", encoding="utf-8") as f:
                page = f.read()

            updated = _replace_edit_id(page, edit_id, text)

            with open(target, "w", encoding="utf-8") as f:
                f.write(updated)

            self._send(200, b"ok")
        except KeyError as e:
            self._send(400, str(e).encode("utf-8"))
        except Exception as e:
            self._send(500, str(e).encode("utf-8"))


def main():
    port = int(os.environ.get("PORT", str(DEFAULT_PORT)))
    server = ThreadingHTTPServer(("127.0.0.1", port), Handler)
    print(f"Edit server running on http://127.0.0.1:{port}")
    print("Open /scenarios.html, then double-click StoryCall text to edit.")
    server.serve_forever()


if __name__ == "__main__":
    main()

File: edit_server.js
#!/usr/bin/env node
/**
 * Local edit server for scenarios.html (and other static files).
 * - Serves files from repo root
 * - Accepts POST /__save { file, id, text } and rewrites the element with data-edit-id="id"
 *
 * Run:
 *   node edit_server.js
 * Then open:
 *   http://127.0.0.1:8787/scenarios.html
 */

const http = require("http");
const fs = require("fs");
const path = require("path");

const ROOT = path.resolve(__dirname);
const PORT = Number(process.env.PORT || 8787);

function safeJoin(root, rel) {
  const clean = String(rel || "").replace(/\\/g, "/").replace(/^\/+/, "");
  if (clean.split("/").includes("..")) throw new Error("bad path");
  const out = path.resolve(root, clean);
  if (!out.startsWith(root + path.sep)) throw new Error("bad path");
  return out;
}

function escapeHtmlText(s) {
  return String(s)
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
}

function replaceEditId(htmlText, editId, newText) {
  const escaped = escapeHtmlText(newText);
  const re = new RegExp(
    `(<(?<tag>[A-Za-z0-9]+)(?<attrs>[^>]*\\sdata-edit-id="${editId.replace(
      /[.*+?^${}()|[\]\\]/g,
      "\\$&"
    )}"[^>]*)>)(?<inner>[\\s\\S]*?)(</\\k<tag>>)`
  );

  const m = htmlText.match(re);
  if (!m || !m.groups) throw new Error(`edit id not found: ${editId}`);

  const start = m.index + m[1].length;
  const end = start + m.groups.inner.length;
  return htmlText.slice(0, start) + escaped + htmlText.slice(end);
}

function contentTypeFor(filePath) {
  if (filePath.endsWith(".html")) return "text/html; charset=utf-8";
  if (filePath.endsWith(".css")) return "text/css; charset=utf-8";
  if (filePath.endsWith(".js")) return "text/javascript; charset=utf-8";
  if (filePath.endsWith(".png")) return "image/png";
  if (filePath.endsWith(".jpg") || filePath.endsWith(".jpeg")) return "image/jpeg";
  if (filePath.endsWith(".webp")) return "image/webp";
  return "application/octet-stream";
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    let buf = "";
    req.setEncoding("utf8");
    req.on("data", (chunk) => (buf += chunk));
    req.on("end", () => resolve(buf));
    req.on("error", reject);
  });
}

const server = http.createServer(async (req, res) => {
  const url = new URL(req.url, "http://127.0.0.1");

  if (req.method === "POST" && url.pathname === "/__save") {
    try {
      const raw = await readBody(req);
      const payload = JSON.parse(raw || "{}");
      const file = payload.file;
      const id = payload.id;
      const text = payload.text ?? "";
      if (!file || !id) {
        res.writeHead(400, { "Content-Type": "text/plain; charset=utf-8" });
        res.end("Missing file or id");
        return;
      }

      const target = safeJoin(ROOT, file);
      const page = fs.readFileSync(target, "utf8");
      const updated = replaceEditId(page, id, text);
      fs.writeFileSync(target, updated, "utf8");

      res.writeHead(200, { "Content-Type": "text/plain; charset=utf-8" });
      res.end("ok");
      return;
    } catch (e) {
      res.writeHead(500, { "Content-Type": "text/plain; charset=utf-8" });
      res.end(String(e && e.stack ? e.stack : e));
      return;
    }
  }

  if (req.method !== "GET" && req.method !== "HEAD") {
    res.writeHead(405, { "Content-Type": "text/plain; charset=utf-8" });
    res.end("Method not allowed");
    return;
  }

  let pathname = url.pathname;
  if (pathname === "/") pathname = "/scenarios.html";

  try {
    const filePath = safeJoin(ROOT, pathname);
    if (!fs.existsSync(filePath) || !fs.statSync(filePath).isFile()) {
      res.writeHead(404, { "Content-Type": "text/plain; charset=utf-8" });
      res.end("Not found");
      return;
    }

    const data = fs.readFileSync(filePath);
    res.writeHead(200, { "Content-Type": contentTypeFor(filePath) });
    if (req.method === "HEAD") {
      res.end();
    } else {
      res.end(data);
    }
  } catch (e) {
    res.writeHead(500, { "Content-Type": "text/plain; charset=utf-8" });
    res.end(String(e && e.stack ? e.stack : e));
  }
});

server.listen(PORT, "127.0.0.1", () => {
  console.log(`Edit server running: http://127.0.0.1:${PORT}`);
  console.log(`Open: http://127.0.0.1:${PORT}/scenarios.html`);
});

